server/src/main/java/io/crate/protocols/postgres/PgDecoder.java

```diff
@@ -96,36 +111,70 @@ protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) t
         }
     }
 
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        super.channelRead(ctx, msg);
+    }
+
     private ByteBuf decode(ChannelHandlerContext ctx, ByteBuf in) {
         switch (state) {
             case STARTUP: {
-                if (in.readableBytes() < MIN_STARTUP_LENGTH) {
+                LOGGER.trace("Readable bytes: {}", in.readableBytes());
+                if (in.readableBytes() == 0) {
+                    // No data at all, don't bother with any complaints logged as such cases often occur for legitimate
+                    // reasons. E.g. some monitoring solutions may open and close the connections without sending any data.
+                    LOGGER.trace("No data received, flush output buffer");
+                    Channel channel = ctx.channel();
+                    ByteBuf out = ctx.alloc().buffer(1);
+                    channel.writeAndFlush(out.writeByte('0'));
                     return null;
                 }
 
                 in.markReaderIndex();
-                payloadLength = in.readInt() - 8;
+                payloadLength = in.readInt();
+
+                if (payloadLength < MIN_STARTUP_LENGTH || payloadLength > MAX_STARTUP_LENGTH) {
+                    in.skipBytes(in.readableBytes());
+                    throw new IllegalStateException("invalid length of startup packet");
+                }
+                payloadLength -= 8; // length includes itself and the request code
+
                 int requestCode = in.readInt();
 
-                if (requestCode == SSL_REQUEST_CODE) {
-                    SslContext sslContext = getSslContext.get();
-                    Channel channel = ctx.channel();
-                    ByteBuf out = ctx.alloc().buffer(1);
-                    if (sslContext == null) {
-                        channel.writeAndFlush(out.writeByte('N'));
-                    } else {
-                        channel.writeAndFlush(out.writeByte('S'));
-                        ctx.pipeline().addFirst(sslContext.newHandler(ctx.alloc()));
-                    }
-                    in.markReaderIndex();
-                    return decode(ctx, in);
-                } else {
-                    if (in.readableBytes() < payloadLength) {
-                        in.resetReaderIndex();
-                        return null;
-                    }
-                    state = requestCode == CANCEL_CODE ? State.CANCEL : State.STARTUP_PARAMETERS;
-                    return in.readBytes(payloadLength);
+                LOGGER.trace("Received startup message with code: {}", requestCode);
+                LOGGER.trace("Payload length: {}", payloadLength);
```

Comment:
```suggestion
```

Already logged above? (except the -8, but that shouldn't be necessary to log again?)

Reply:
No it isn't logged above, only the readable bytes are logged. But given that these trace logs are mainly for debugging purpose, there is no real need in committing this. wdyt?

Reply:
Oh my bad. Maybe merge the two entries into one? `Received startup message code={} length={}` ?